pmlc/dialect/pxa/tests/ops.mlir

```diff
@@ -0,0 +1,159 @@
+  %a = alloc() : memref<100x100xf32>
+  %r = affine.parallel (%i, %j, %k) = (0, 0, 0) to (100, 100, 100) reduce ("assign") -> (memref<100x100xf32>) {
+    %0 = affine.vector_load %arg1[%i, %k] : memref<100x100xf32>, vector<4xf32>
+    %1 = affine.vector_load %arg0[%k, %j] : memref<100x100xf32>, vector<4xf32>
+    %2 = mulf %0, %1 : vector <4xf32>
+    %red = pxa.vector_reduce assign %2, %a[%i, %j] :  memref<100x100xf32>, vector<4xf32>
+    // CHECK: %{{.*}} = affine.vector_load %{{.*}}[] : memref<100x100xf32>, vector<4xf32>
+    // CHECK: affine.vector_store %{{.*}}, %{{.*}}[] : memref<100x100xf32>, vector<4xf32>
+    affine.yield %red : memref<100x100xf32>
+  }
+  return %r : memref<100x100xf32>
+}
+
+// CHECK-LABEL: func @pxa_reduce_add
+func @pxa_reduce_add(%arg0: memref<100x100xf32>, %arg1: memref<100x100xf32>) -> (memref<100x100xf32>) {
+  %a = alloc() : memref<100x100xf32>
+  %r = affine.parallel (%i, %j, %k) = (0, 0, 0) to (100, 100, 100) reduce ("assign") -> (memref<100x100xf32>) {
+    %0 = affine.load %arg1[%i, %k] : memref<100x100xf32>
+    %1 = affine.load %arg0[%k, %j] : memref<100x100xf32>
+    %2 = mulf %0, %1 : f32
+    %red = pxa.reduce add %2, %a[%i, %j] :  memref<100x100xf32>
+    // CHECK: %{{.*}} = affine.load %{{.*}} : memref<100x100xf32>
+    // CHECK: %{{.*}} = addf %{{.*}}, %{{.*}} : f32
+    // CHECK: affine.store %{{.*}}, %{{.*}} : memref<100x100xf32>
+    affine.yield %red : memref<100x100xf32>
+  }
+  return %r : memref<100x100xf32>
+}
+
+// CHECK-LABEL: func @pxa_vector_reduce_add
+func @pxa_vector_reduce_add(%arg0: memref<100x100xf32>, %arg1: memref<100x100xf32>) -> (memref<100x100xf32>) {
+  %a = alloc() : memref<100x100xf32>
+  %r = affine.parallel (%i, %j, %k) = (0, 0, 0) to (100, 100, 100) reduce ("assign") -> (memref<100x100xf32>) {
+    %0 = affine.vector_load %arg1[%i, %k] : memref<100x100xf32>, vector<4xf32>
+    %1 = affine.vector_load %arg0[%k, %j] : memref<100x100xf32>, vector<4xf32>
+    %2 = mulf %0, %1 : vector <4xf32>
+    %red = pxa.vector_reduce add %2, %a[%i, %j] :  memref<100x100xf32>, vector<4xf32>
+    // CHECK: %{{.*}} = affine.vector_load %{{.*}}[] : memref<100x100xf32>, vector<4xf32>
+    // CHECK: %{{.*}} = addf %{{.*}}, %{{.*}} : vector<4xf32>
+    // CHECK: affine.vector_store %{{.*}}, %{{.*}}[] : memref<100x100xf32>, vector<4xf32>
+    affine.yield %red : memref<100x100xf32>
+  }
+  return %r : memref<100x100xf32>
+}
+
+// CHECK-LABEL: func @pxa_reduce_mul
+func @pxa_reduce_mul(%arg0: memref<100x100xf32>, %arg1: memref<100x100xf32>) -> (memref<100x100xf32>) {
+  %a = alloc() : memref<100x100xf32>
+  %r = affine.parallel (%i, %j, %k) = (0, 0, 0) to (100, 100, 100) reduce ("assign") -> (memref<100x100xf32>) {
+    %0 = affine.load %arg1[%i, %k] : memref<100x100xf32>
+    %1 = affine.load %arg0[%k, %j] : memref<100x100xf32>
+    %2 = mulf %0, %1 : f32
+    %red = pxa.reduce mul %2, %a[%i, %j] :  memref<100x100xf32>
+    // CHECK: %{{.*}} = affine.load %{{.*}} : memref<100x100xf32>
```

Comment:
The patterns are a bit needlessly permissive. You should probably do captures of the values.  You could verify that for example that the loaded value is passed into the mulf and that the output of the mulf goes into the store, and that the store and the load both reference the same buffer (which they should post pxa->affine)

Reply:
Added more soecific checks.